src/ocp_postprocess/ip_rename/etcd_rename.rs

```diff
@@ -6,10 +6,77 @@ use crate::{
 use anyhow::{bail, ensure, Context, Result};
 use futures_util::future::join_all;
 use serde_json::Value;
-use std::net::Ipv6Addr;
+use std::net::{IpAddr, Ipv6Addr};
 use std::sync::Arc;
 
-pub(crate) async fn fix_openshift_apiserver_configmap(etcd_client: &Arc<InMemoryK8sEtcd>, ip: &str) -> Result<String> {
+// Helper function to determine if an IP is IPv4 or IPv6
+fn is_ipv6(ip: &str) -> bool {
+    ip.parse::<IpAddr>().map(|addr| addr.is_ipv6()).unwrap_or(false)
+}
+
+// Extract both original IPv4 and IPv6 IPs from dual-stack cluster node configuration
+pub(crate) async fn extract_original_ips(etcd_client: &Arc<InMemoryK8sEtcd>) -> Result<Vec<String>> {
+    // Extract IPs from node addresses - works for both single-stack and dual-stack
+    extract_original_ips_from_nodes(etcd_client).await
+}
+
+// Extract original IPs from node configuration - returns 1 IP for single-stack, 2 for dual-stack
+async fn extract_original_ips_from_nodes(etcd_client: &Arc<InMemoryK8sEtcd>) -> Result<Vec<String>> {
+    let node_keys = etcd_client.list_keys("minions").await?;
+    ensure!(!node_keys.is_empty(), "No nodes found in etcd");
+
+    // Get the first (and should be only) node
+    let node_key = &node_keys[0];
```

Comment:
Better to encode assumptions in assertions than comments

```suggestion
    let node_keys = etcd_client.list_keys("minions").await?;

    ensure!(
        node_keys.len() == 1,
        "Expected exactly one node in the cluster, found {}",
        node_keys.len()
    );

    let node_key = &node_keys[0];
```